node/sync/src/ping.rs

```diff
@@ -0,0 +1,198 @@
+/// for when a peer should next be pinged.
+///
+/// TODO (kaimast): maybe keep track of the last ping too, to not trigger spam detection?
+struct PingInner<N: Network> {
+    /// The next time we should ping a peer.
+    next_ping: BTreeMap<Instant, SocketAddr>,
+    /// The most recent block locators.
+    /// (or None if this node does not offer block sync)
+    block_locators: Option<BlockLocators<N>>,
+}
+
+/// Manages sending Ping messages to all connected peers.
+pub struct Ping<N: Network> {
+    router: Router<N>,
+    inner: Arc<Mutex<PingInner<N>>>,
+    notify: Arc<Notify>,
+}
+
+impl<N: Network> PingInner<N> {
+    fn new(block_locators: Option<BlockLocators<N>>) -> Self {
+        Self { block_locators, next_ping: Default::default() }
+    }
+}
+
+impl<N: Network> Ping<N> {
+    /// The duration in seconds to wait between sending ping requests to a peer.
+    const MAX_PING_INTERVAL: Duration = Duration::from_secs(20);
+
+    /// Create a new instance of the ping logic.
+    /// There should only be one per node.
+    ///
+    /// # Usage
+    /// Initialize this with the most up-to-date block locators and call
+    /// update_block_locators, whenever a new block is received/created.
+    pub fn new(router: Router<N>, block_locators: BlockLocators<N>) -> Self {
+        let notify = Arc::new(Notify::default());
+        let inner = Arc::new(Mutex::new(PingInner::new(Some(block_locators))));
+
+        {
+            let inner = inner.clone();
+            let router = router.clone();
+            let notify = notify.clone();
+
+            tokio::spawn(async move {
+                Self::ping_task(&inner, &router, &notify).await;
+            });
+        }
+
+        Self { inner, router, notify }
+    }
+
+    /// Same as [`Self::new`] but for nodes that peers cannot sync from
+    /// such as provers.
+    pub fn new_nosync(router: Router<N>) -> Self {
+        let notify = Arc::new(Notify::default());
+        let inner = Arc::new(Mutex::new(PingInner::new(None)));
+
+        {
+            let inner = inner.clone();
+            let router = router.clone();
+            let notify = notify.clone();
+
+            tokio::spawn(async move {
+                Self::ping_task(&inner, &router, &notify).await;
+            });
+        }
+
+        Self { inner, router, notify }
+    }
+
+    /// Notify the ping logic that we received a Pong response.
+    pub fn on_pong_received(&self, peer_ip: SocketAddr) {
+        let now = Instant::now();
+        let mut inner = self.inner.lock();
+
+        inner.next_ping.insert(now + Self::MAX_PING_INTERVAL, peer_ip);
+
+        // self.notify.notify() is not needed as ping_task wakes up every MAX_PING_INTERVAL
+    }
+
+    /// Notify the ping logic that a new peer connected.
+    pub fn on_peer_connected(&self, peer_ip: SocketAddr) {
+        // Send the first ping.
+        let locators = self.inner.lock().block_locators.clone();
+        self.router.send_ping(peer_ip, locators);
```

Comment:
Should this attempt to reset any respective entry in `inner.next_ping`, i.e. set the max ping interval after pinging?